stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java

```diff
@@ -313,13 +319,83 @@ private BStruct createUserDefinedType(Struct structValue, BStructType structType
 
     private void handleArrayValue(BStruct bStruct, RegistryIndex refRegIndex, int index) throws SQLException {
         BStructType.StructField[] structFields = getStructFields();
-        int fieldTypeTag = getFieldTypeTag(structFields, index);
-        Array dataArray = rs.getArray(index);
-        if (fieldTypeTag == TypeTags.UNION_TAG) {
-            validateAndSetRefRecordField(bStruct, refRegIndex.incrementAndGet(), TypeTags.ARRAY_TAG,
-                    retrieveType(structFields, index - 1), getDataArray(dataArray));
+        BType fieldType = structFields[index - 1].getFieldType();
+        int fieldTypeTag = fieldType.getTag();
+        Array data = rs.getArray(index);
+        BNewArray dataArray = getDataArray(data);
+        if (dataArray != null) {
+            BType nonNilType = fieldType;
+            if (fieldTypeTag == TypeTags.UNION_TAG) {
+                nonNilType = retrieveNonNilType(((BUnionType) structFields[index - 1].getFieldType()).getMemberTypes());
+            }
+            handleMappingArrayValue(nonNilType, bStruct, dataArray, refRegIndex,
+                    dataArray.getType().getTag() == TypeTags.UNION_TAG);
+        } else {
+            if (fieldTypeTag == TypeTags.UNION_TAG) {
+                bStruct.setRefField(refRegIndex.incrementAndGet(), null);
+            } else {
+                handleNilToNonNillableFieldAssignment();
+            }
+        }
+    }
+
+    private void handleMappingArrayValue(BType nonNilType, BStruct bStruct, BNewArray dataArray,
+            RegistryIndex refRegIndex, boolean containsNull) {
+        if (nonNilType.getTag() == TypeTags.ARRAY_TAG) {
+            BArrayType expectedArrayType = (BArrayType) nonNilType;
+            if (((BArrayType) nonNilType).getElementType().getTag() == TypeTags.UNION_TAG) {
+                handleMappingArrayElementToUnionType(expectedArrayType, dataArray, refRegIndex, bStruct);
+            } else {
+                handleMappingArrayElementToNonUnionType(containsNull, nonNilType, dataArray, bStruct, refRegIndex);
+            }
         } else {
-            bStruct.setRefField(refRegIndex.incrementAndGet(), getDataArray(dataArray));
+            handleMismatchingFieldAssignment();
+        }
+    }
+
+    private void handleMappingArrayElementToNonUnionType(boolean containsNull, BType nonNilType, BNewArray newArray,
+            BStruct bStruct, RegistryIndex refRegIndex) {
+        if (containsNull) {
+            throw new BallerinaException(
+                    "Trying to assign an array containing NULL values to an array of a non-nillable " + "element type");
+        } else {
+            validateAndSetNonUnionRefRecordField(bStruct, refRegIndex.incrementAndGet(), ((BArrayType) nonNilType)
```

Comment:
Let's call validateAndSetRefRecordField directly here without having two level of indirection which is not reused